File: LeetCode/0743-network-delay-time/0743-network-delay-time.java
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        Map<Integer, Map<Integer, Integer>> graph = new HashMap<>();
        
        for (int[] time : times) {
            graph.putIfAbsent(time[0], new HashMap<>());
            graph.get(time[0]).put(time[1], time[2]);
        }
        
        Queue<int[]> pq = new PriorityQueue<>((arr1, arr2) -> arr1[1] - arr2[1]);
        
        pq.add(new int[] {k, 0});
        
        Map<Integer, Integer> dist = new HashMap<>();
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int node = curr[0];
            int cost = curr[1];
            
            if (!dist.containsKey(node)) {
                dist.put(node, cost);
                
                if (dist.get(node) < cost) continue;
                
                if (graph.containsKey(node)) {
                    for (int next : graph.get(node).keySet()) {
                        int alt = cost + graph.get(node).get(next);
                        
                        pq.add(new int[] {next, alt});
                    }
                }
            }
        }
        
        if (dist.size() == n) {
            return Collections.max(dist.values());
        }
        
        return -1;
    }
}

File: LeetCode/0743-network-delay-time/0743-network-delay-time.kt
class Solution {
    fun networkDelayTime(times: Array<IntArray>, n: Int, k: Int): Int {
        fun initGraph(): Map<Int, List<Destination>> {
            val graph: MutableMap<Int, MutableList<Destination>> = mutableMapOf()
            
            for (time in times) {
                graph.computeIfAbsent(time[0]){ mutableListOf() }
                    .add(Destination(time[1], time[2]))
            }
            
            return graph
        }
        
        val graph: Map<Int, List<Destination>> = initGraph()
        val distances: MutableMap<Int, Int> = mutableMapOf()
        val queue: Queue<Destination> = PriorityQueue(Comparator.comparingInt({ d -> d.distance }))
        
        queue.add(Destination(k, 0))
        
        fun updateDistance() {
            while(!queue.isEmpty()) {
                val destination: Destination = queue.poll()
                val node: Int = destination.node
                val distance: Int = destination.distance
                
                if (distances.containsKey(node)) {
                    continue
                }
                
                distances[node] = distance
                
                if (!graph.containsKey(node)) {
                    continue
                }
                
                for (nextDestination in graph[node]!!) {
                    val nextDistance: Int = distance + nextDestination.distance
                    
                    queue.add(Destination(nextDestination.node, nextDistance))
                }
            }
        }
        
        updateDistance()
        
        return if (distances.size != n) -1 else Collections.max(distances.values)
    }
    
    class Destination(val node: Int, val distance: Int) {
    }
}